File: project_euler_unified/problem_049/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int LIMIT = 10000;
    vector<bool> is_prime(LIMIT, true);
    is_prime[0] = is_prime[1] = false;
    for (int i = 2; i * i < LIMIT; i++)
        if (is_prime[i])
            for (int j = i * i; j < LIMIT; j += i)
                is_prime[j] = false;

    map<string, vector<int>> groups;
    for (int p = 1000; p < LIMIT; p++) {
        if (!is_prime[p]) continue;
        string s = to_string(p);
        sort(s.begin(), s.end());
        groups[s].push_back(p);
    }

    for (auto& [key, primes] : groups) {
        if ((int)primes.size() < 3) continue;
        set<int> pset(primes.begin(), primes.end());
        for (int i = 0; i < (int)primes.size(); i++) {
            for (int j = i + 1; j < (int)primes.size(); j++) {
                int third = 2 * primes[j] - primes[i];
                if (third < LIMIT && pset.count(third) && primes[i] != 1487) {
                    cout << primes[i] << primes[j] << third << endl;
                    return 0;
                }
            }
        }
    }
    return 0;
}

File: project_euler_unified/problem_049/solution.py
"""
Problem 49: Prime Permutations

Find the other 3-term arithmetic sequence of 4-digit primes that are
permutations of each other, and concatenate the three terms.
"""

from collections import defaultdict

def sieve(limit):
    is_prime = [True] * limit
    is_prime[0] = is_prime[1] = False
    for i in range(2, int(limit**0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, limit, i):
                is_prime[j] = False
    return is_prime

def solve():
    is_prime = sieve(10000)

    groups = defaultdict(list)
    for p in range(1000, 10000):
        if is_prime[p]:
            groups[''.join(sorted(str(p)))].append(p)

    for primes in groups.values():
        if len(primes) < 3:
            continue
        pset = set(primes)
        for i in range(len(primes)):
            for j in range(i + 1, len(primes)):
                third = 2 * primes[j] - primes[i]
                if third in pset and primes[i] != 1487:
                    return f"{primes[i]}{primes[j]}{third}"

print(solve())
